File: backend/app/schemas/problem.py
from uuid import UUID

from pydantic import BaseModel


class ProblemResponse(BaseModel):
    id: UUID
    source: str | None
    subject: str
    format: str
    question_markdown: str
    options: dict | None
    difficulty: int | None
    estimated_time_sec: int | None
    hints: list[str] | None = None
    correct_answer: str | None = None
    knowledge_point_ids: list[str] | None = None
    explanation: str | None = None

    model_config = {"from_attributes": True}


class AttemptRequest(BaseModel):
    answer: str
    session_id: UUID | None = None
    time_spent_sec: int | None = None
    hint_level: int = 0


class AttemptResponse(BaseModel):
    id: UUID
    is_correct: bool | None
    ai_feedback: str | None
    error_type: str | None = None
    attempt_number: int
    xp_earned: int = 0

    model_config = {"from_attributes": True}

File: frontend/src/types/problem.ts
// Types derived from backend Pydantic schemas:
//   backend/app/schemas/problem.py
//   backend/app/schemas/chat.py
//   backend/app/models/problem.py
//   backend/app/models/message.py

// ── API Response Types (match backend exactly) ──────────────────────────

/** Backend: ProblemResponse – GET /problems/:id, GET /practice/next */
export interface ProblemResponse {
  id: string;
  source: string | null;
  subject: string;
  format: string;
  question_markdown: string;
  options: Record<string, unknown> | null;
  difficulty: number | null;
  estimated_time_sec: number | null;
}

/** Backend: AttemptRequest – POST /problems/:id/attempts */
export interface AttemptRequest {
  answer: string;
  session_id?: string | null;
  time_spent_sec?: number | null;
}

/** Backend: AttemptResponse – POST /problems/:id/attempts response */
export interface AttemptResponse {
  id: string;
  is_correct: boolean | null;
  ai_feedback: string | null;
  error_type?: string | null;
  attempt_number: number;
  xp_earned: number;
}

/** Backend: ChatMessageResponse – chat messages */
export interface ChatMessageResponse {
  id: string;
  role: string;
  content: string;
  session_id: string | null;
}

// ── Backward-compatible aliases / extended types ─────────────────────────

export interface ChatMessage extends Omit<ChatMessageResponse, 'session_id'> {
  session_id?: string | null;
  timestamp?: string;
  metadata?: {
    latex?: string;
    image_url?: string;
    audio_url?: string;
  };
}

// ── Frontend-only types ─────────────────────────────────────────────────

/** Backend: ChatMessageRequest – POST /chat */
export interface ChatMessageRequest {
  session_id?: string | null;
  content: string;
  media?: Record<string, unknown> | null;
}

/** UI-only problem type used by practice-session component */
export interface Problem {
  id: string;
  type: string;
  content: string;
  latexContent?: string;
  options?: string[];
  correctAnswer: string;
  explanation: string;
  hintLevels: string[];
  xpReward: number;
}

export interface KnowledgeState {
  knowledge_point_id: string;
  name: string;
  p_correct: number;
  attempts: number;
  last_practiced?: string;
}
